exporter/awsemfexporter/internal/useragent/useragent.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package useragent // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter/internal/useragent"
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/jellydator/ttlcache/v3"
+	semconv "go.opentelemetry.io/collector/semconv/v1.22.0"
+)
+
+const (
+	handlerName = "aws.awsemfexporter.UserAgentHandler"
+	// defaultTTL is how long an item in the cache will remain if it has not been re-seen.
+	defaultTTL = time.Minute
+	// cacheSize is the maximum number of unique telemetry SDK languages that can be stored before one will be evicted.
+	cacheSize = 10
+	// attrLengthLimit is the maximum length of the language and version that will be used for the user agent.
+	attrLengthLimit = 20
+	// Relic attribute from before semconv/v1.21.0. Replaced with "telemetry.distro.version", but old SDKs may still be using it
+	attributeTelemetryAutoVersion = "telemetry.auto.version"
+)
+
+type UserAgent struct {
+	mu          sync.RWMutex
+	prebuiltStr string
+	cache       *ttlcache.Cache[string, map[string]struct{}]
+}
+
+func NewUserAgent() *UserAgent {
+	return newUserAgent(defaultTTL)
+}
+
+func newUserAgent(ttl time.Duration) *UserAgent {
+	ua := &UserAgent{
+		cache: ttlcache.New[string, map[string]struct{}](
+			ttlcache.WithTTL[string, map[string]struct{}](ttl),
+			ttlcache.WithCapacity[string, map[string]struct{}](cacheSize),
+		),
+	}
+	ua.cache.OnEviction(func(context.Context, ttlcache.EvictionReason, *ttlcache.Item[string, map[string]struct{}]) {
+		ua.build()
+	})
+	go ua.cache.Start()
+	return ua
+}
+
+// Handler creates a named handler with the UserAgent's handle function.
+func (ua *UserAgent) Handler() request.NamedHandler {
+	return request.NamedHandler{
+		Name: handlerName,
+		Fn:   ua.handle,
+	}
+}
+
+// handle adds the pre-built user agent string to the user agent header.
+func (ua *UserAgent) handle(r *request.Request) {
+	ua.mu.RLock()
+	defer ua.mu.RUnlock()
+	request.AddToUserAgent(r, ua.prebuiltStr)
+}
+
+// Process takes the telemetry SDK language and version and adds them to the cache. If the language already exists in the
+// cache with a different version, then add the new version to the version map. If not, then create a new version map.
+func (ua *UserAgent) Process(labels map[string]string) {
+	language := labels[semconv.AttributeTelemetrySDKLanguage]
+	version := labels[semconv.AttributeTelemetryDistroVersion]
+	if version == "" {
+		version = labels[attributeTelemetryAutoVersion]
+	}
+	if language != "" && version != "" {
+		language = truncate(language, attrLengthLimit)
+		version = truncate(version, attrLengthLimit)
+		value := ua.cache.Get(language)
+		valueMap := make(map[string]struct{})
```

Comment:
The name `valueMap` may be a bit confusing, given that we are using this value to store a set of versions.
Can we update the variable name (e.g. `versionSet`) and  add comment that we are treating the map as a set?